### ISSUE TITLE
bug: KS v0.21 latency for remote pod execution degraded relative to v0.20 for Braulio stress test

### ISSUE THREAD

**Issue by eaepstein** [kubestellar/kubestellar#1954]
### Describe the bug

Braulio's stress test creates pods in a WDS at a specified rate. The pods run "bash -c sleep 20" on the ubuntu:latest image and then terminate. In my test KS was deployed on a VM (32GB RAM, 8vCPU) using K3D clusters for hub and spoke. 

The two plots below show individual pod end-to-end latencies from creation time in WDS to status=complete in WDS. Pods were created 1 every 3 seconds in the same namespace. A single binding policy was used. There were no pod deletions during each test.

![k3d-ks020-end-to-end-onethird](https://github.com/kubestellar/kubestellar/assets/56738503/ad34f71f-8df9-4b14-9c53-82889e414a73)

![k3d-ks021-end-to-end-onethird](https://github.com/kubestellar/kubestellar/assets/56738503/5435c686-d8f4-4f0c-8ba7-8e202a6de607)

### Steps To Reproduce

A PR adding a document describing how to deploy KS on K3D should be available shortly.
Braulio will document how to run his stress test. Yaml files for the pod and and the binding policy are included here.

[yamls.zip](https://github.com/kubestellar/kubestellar/files/14727649/yamls.zip)

### Expected Behavior

KS overhead latency should not be so badly degraded for this test.

**Comment by nirrozenbaum** [comment]
@eaepstein @dumb0002 thanks for opening this issue, indeed an important one that worth deeper investigation.
would it be possible to add more information about where do we spend the time?
it would be very helpful to understand what exactly takes time to identify where is the problem.
some examples - 
- does the distribution from WDS to ITS take more time?
- does OCM distribution from ITS to WECs take more time? in this case we might hit OCM bug (see issue #1868)
- it could be that distribution works perfectly fine but the issue is in the status?
- etc...

**Comment by ezrasilvera** [comment]
@nirrozenbaum I don't think it's an issue of "where we spend the time" , the main problem I see is a (linear) constant increase in time as the number of resources  increases.  Which means this is a major scalability issue.

**Comment by ezrasilvera** [comment]
@eaepstein @nirrozenbaum One test that can be helpful is:
1. Measure latency of creation
2. Populate the system with large number of objects
3. Wait for few minutes 
4. Repeat (1) - Measure latency of creation 
If 4 is >> from 1 we have  a scaling issue that can be easily profiled and debugged separately (out of the "stress test" environment)

**Comment by nirrozenbaum** [comment]
@ezrasilvera before you declare scalability issue, I would first try to analyze things in a more fine-grained granularity.
we may find a bug that's causing this and upon solving the bug we may find out things are working as expected.
we may also find out the issue is related to OCM and not to KS.
therefore, it's important to understand where we spend the time and what takes so long.
it should be easy to get this information.

**Comment by effi-ofer** [comment]
This is a very interesting issue.  It would be interesting to see the time spent in each phase (object creation in WDS, binding object creation, manifestwork object creation, object creation in WEC).  I am hoping this can be easily obtained from the creationTimestamp field in the different objects.

**Comment by ezrasilvera** [comment]
@nirrozenbaum I'm not sure I follow your point - The test I was proposing is the first step to such analysis .   If you see consistent increase in  the test I proposed you can easily profile it and see exactly where the times were spent. 
If you don't see that increase , it means that it also related to the dynamic of the test (rate of creation, creation+status delivery, etc..) and this is much harder to profile.

**Comment by ezrasilvera** [comment]
Also - a simple test would be to repeat the exact same test without the "Want Status"  .   While continue getting similar results won't prove that's it's still not related to the status, getting improved results can prove that the latency is related to status.

**Comment by nirrozenbaum** [comment]
@ezrasilvera ok, you're saying what I was saying in different words.
I was asking to see where do we spend time, while you are suggesting ways to check it.
so we're basically requesting for the same using different words :).

**Comment by ezrasilvera** [comment]
@dumb0002 (cc: @eaepstein )  Can you also attach or point to the script you use to test and to measure the timing (in case we need to reproduce)

**Comment by dumb0002** [comment]
@nirrozenbaum @ezrasilvera @effi-ofer (cc; @eaepstein) - sure, I will investigate where we spend the time in each phase to identify any bottleneck phase in the overall e2e latency illustrated in the graph above.  @ezrasilvera, here a draft README.md on how to run the scripts: https://github.ibm.com/MCC/MCC/tree/mcc-exp-v2/exp

**Comment by ezrasilvera** [comment]
@dumb0002 I think it's good that KS would have a performance section. I understand that currently your tests are in the MCC scope but maybe they can be easily updated to run directly on top of KS. 
I created https://github.com/kubestellar/kubestellar/issues/1960   to track that.   
We also need to find a good place to store the results under the KS git

**Comment by eaepstein** [comment]
@nirrozenbaum Braulio's data collection was designed for separate manifestwork objects for each workload object, so the standard data analysis is broken for wds-pod-create -> imbs-manifestwork-updated. Does each entry in the bundled manifestwork have a create time? Do entries also have the ID of the associated workload objects to be able to measure transport times?

The plot below measures from wds-pod-create -> wec-pod-create, so it includes OCM time. But it also makes it clear that the end-to-end latency degradation is due to the effect of bundling, as nothing else changed.

![ks021-transport-vs-end2end](https://github.com/kubestellar/kubestellar/assets/56738503/9478b55f-d731-47c3-88b1-5cf9efc186fc)

**Comment by dumb0002** [comment]
@ezrasilvera, my current stress test can run directly on top of KS without changes, I'll improve the readme and move it to the KS repo (using  #1960 to keep track of it)

**Comment by nirrozenbaum** [comment]
thanks @eaepstein and @dumb0002.
if we keep creating pods at the same rate, do we get to a stable graph after a while? or is it continuously keep growing?
is it possible to run this test for 30 minutes (or even more) and show the output?

**Comment by eaepstein** [comment]
@nirrozenbaum The longer run of ks0.21 stopped down-syncing at 187 pods. Latencies were continuing to go up.
There was nothing notable in the klusterlet or transport-ctlr logs. At some point as the existing pods were being deleted from WDS, the remaining pending pods began downsyncing again.

A ks0.20 test of 400 pods completed. The latency charts are below. 

![ks020-400-vs-ks121-187](https://github.com/kubestellar/kubestellar/assets/56738503/dc5608d7-55a5-4648-8bd5-f20b0c6dcb0d)

**Comment by ezrasilvera** [comment]
> thanks @eaepstein and @dumb0002. if we keep creating pods at the same rate, do we get to a stable graph after a while? or is it continuously keep growing? is it possible to run this test for 30 minutes (or even more) and show the output?

@nirrozenbaum While it is interesting question, I'm not sure why we need to discuss it or try it now.  I think what we already see  is good enough reason to profile and explore the reason for the linear growth in latency.  The only question is who is going to do it..   I'm not sure that the person who is going to profile and analyze it needs more info from @dumb0002 ..
We simply need someone to take this issue and investigate.

**Comment by nirrozenbaum** [comment]
@ezrasilvera what I was trying to understand is if it reaches some stability. if yes, there is no scale issue, cause we may see the scale continuously growing (the pod creation rate is faster than completion rate) and there is no increase in time.
I think this is a very important input.

it could be that in this specific setup (which I was continuously describing as not representative for KS), in the warmup phase of the test the measured time is slowly growing but then reaches stability.
if that would be the case I would be very happy to see that. that's a very good sign. 

of course we can assign a person to further investigate but there's a major difference whether the issue is in the warmup or a general issue and it keeps growing and growing.

**Comment by nirrozenbaum** [comment]
> @nirrozenbaum The longer run of ks0.21 stopped down-syncing at 187 pods. Latencies were continuing to go up. There was nothing notable in the klusterlet or transport-ctlr logs. At some point as the existing pods were being deleted from WDS, the remaining pending pods began downsyncing again.
> 
> A ks0.20 test of 400 pods completed. The latency charts are below.
> 
> ![ks020-400-vs-ks121-187](https://private-user-images.githubusercontent.com/56738503/316999418-dc5608d7-55a5-4648-8bd5-f20b0c6dcb0d.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1MTkwMzEsIm5iZiI6MTcxMTUxODczMSwicGF0aCI6Ii81NjczODUwMy8zMTY5OTk0MTgtZGM1NjA4ZDctNTVhNS00NjQ4LThiZDUtZjIwYjBjNmRjYjBkLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI3VDA1NTIxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzOWJlNTgyNDQyZjhlZDcyNGQwMGViZmUxMTNlNjBiYTE2M2Y3MmI0M2FhMjVkMDAwMDAxZTQxOWU0YzEzZTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MUZBgEGotdfXcA8ItWxUw4MXzZ4dhuDLULlwujk-THY)

@eaepstein thanks for the input. I just noticed this input AFTER answering Ezra.
@ezrasilvera yes.  it's definitely worth investigating and see where exactly is the problem.

**Comment by nirrozenbaum** [comment]
sharing an initial analysis - 
in ks0.20 each object was wrapped individually. therefore, even though all pods are mapped to the WEC using a single `BindingPolicy`, the result of a new pod every 3 seconds is a new `ManifestWork` that includes one pod every 3 seconds. 
that makes the creation rate + the size of `ManifestWork` stable. we can clearly see in the graph that 0.20 seems to at stable rate + size right from the beginning (the average seems to be somewhere around 24 sec per pod).

in ks0.21 the appraoch we took is different. 
`ManifestWork` includes all objects that were mapped to a WEC using a single `BindingPolicy`. since this test is using a single `BindingPolicy` to map all pods, that means that in each iteration of the test, the `ManifestWork` object keep growing and growing. from an initial analysis it makes sense that the time to sync it to WEC keeps growing (cause the object is larger in every iteration). then at some point it seems not to sync anymore. intuitively I think it could be that the `ManifestWork` object exceeded the limit of a k8s resource size (the limit is 1MB per resource, this is a general limit of k8s not related to `ManifestWork`) since it tries to put more than 187 pods in a single `ManifestWork`.
at some point when pods are starting to go away, the pending pods (all the ones above the 187) are starting to sync again. this assumption has to be further investigated and if correct, we already talked about it in the past to introduce a mechanism for fragmentation (splitting into more than one manifestwork). 
that's an advanced topic so it was not included in the first transport version.

according to this analysis, the graph makes sense to me.

the question then goes back to the original question I asked - 
is this a representative stress test for KS? I think we agreed it's not.
it's not expected from a KS user to use a single `BindingPolicy` to distribute all objects that ever exist, although that's not prohibited. 

I think we need stress the system in a more representative test and look if there are major differences in performance.
If that's ok with all of you, give me around two weeks and I'll try to come up with stress test setup that represents KS usage better. we can then discuss it and continue investigating. 

/cc @vMaroon

**Comment by eaepstein** [comment]
@nirrozenbaum A single binding policy could be in place for multiple users on one WDS, each in their own namespace. The degradation is pretty bad even with low numbers, and appwrapper objects could be much bigger than pods.
It would interesting to find a realistic scenario for which bundling per binding policy has significantly better performance in some metric: latency, memory, or CPU.

**Comment by dumb0002** [comment]
@nirrozenbaum, for the use-case of MCAD +  ks0.20 integration, a decision was made to use a single `BindingPolicy` per remote clusters created at the bootstrap of the env: that user does not directly interacts with KS and uses upper layer application to select where to deploy its workload - This stress test experiment was representing such use-case. 

For ks0.21, it would be interesting to evaluate the performance for both extreme cases: 1) a binding policy per objects  and 2) a single binding policy for all the objects (the current test). In addition to evaluate some cases in between. 

Which more representative use-case for test do you have in mind? Let's have more discussion in this - we can configure the tool that we are using to generate the workload to represent such a scenario - clusterloader2 can be easily configured.

**Comment by nirrozenbaum** [comment]
> @nirrozenbaum A single binding policy could be in place for multiple users on one WDS, each in their own namespace. The degradation is pretty bad even with low numbers, and appwrapper objects could be much bigger than pods. It would interesting to find a realistic scenario for which bundling per binding policy has significantly better performance in some metric: latency, memory, or CPU.

@eaepstein I didn't understand this comment. `BindingPolicy is a cluster-scoped resource. what do you mean when you say single binding policy for multiple users on one WDS, each in their own namespace?
BindingPolicy isn't in a namespace.. can you please explain?

**Comment by nirrozenbaum** [comment]
> @nirrozenbaum, for the use-case of MCAD + ks0.20 integration, a decision was made to use a single `BindingPolicy` per remote clusters created at the bootstrap of the env: that user does not directly interacts with KS and uses upper layer application to select where to deploy its workload - This stress test experiment was representing such use-case.
> 
> For ks0.21, it would be interesting to evaluate the performance for both extreme cases: 1) a binding policy per objects and 2) a single binding policy for all the objects (the current test). In addition to evaluate some cases in between.
> 
> Which more representative use-case for test do you have in mind? Let's have more discussion in this - we can configure the tool that we are using to generate the workload to represent such a scenario - clusterloader2 can be easily configured.

regarding representative test for KS - I think we discussed it several times with the team. if a project has to manage 2 clusters and for each of them they distribute one wrapped object, they probably wouldn't use KS for that :). KS has a big overhead if the purpose is managing only 2 clusters and for each deploying one object.

if we put aside MCAD integration (although we definitely need to support MCAD use case), I would say a representative use case for KS would be something along these lines - 
having 10-100 WECs.
having 10 different users, some of them could share WDS, so let's say we spread them over 3-4 WDS.
each user creates 10 BindingPolicy objects, each selects 5-10 WECs, in each they have 20 objects.

I would expect that a workload that users distributes with a single binding policy to have some relation between the objects. for example - a workload (set of objects) may be depoyment, secret, serviceaccount, role, rolebinding, configmap, etc.. with some dependencies between them. e.g., the deployment uses the service account, the secret and configmap and service account gets its permissions using the role+rolebinding.

again, we do not prohibit distributing in one policy arbitrary set of objects with no relation between them, but that doesn't really makes sense from a user point of view. usually k8s users use labels on objects to map them to some application that runs on the cluster.

so I think we should aim to test the system a setup closer to the above rather than the extreme cases. extreme cases are not the representative cases.

**Comment by dumb0002** [comment]
@nirrozenbaum, thanks for the detailed explanation above. I will start looking at open-source tools to simulate k8s clusters as we'll need simulation to experiment with large number of  WECs.  I still think that to fully understand the performance of system it is important to test for representative cases, as well as the extreme cases - it is important to understand how a system behaves when it is pushed to its limit but we don't have to necessarily fix it for the extreme cases.

**Comment by nirrozenbaum** [comment]
> @nirrozenbaum, thanks for the detailed explanation above. I will start looking at open-source tools to simulate k8s clusters as we'll need simulation to experiment with large number of WECs. I still think that to fully understand the performance of system it is important to test for representative cases, as well as the extreme cases - it is important to understand how a system behaves when it is pushed to its limit but we don't have to necessarily fix it for the extreme cases.

right. I agree.
the performance optimizations should be oriented to the representative cases.
so after seeing a comparison with those cases we can optimize one way or another (depends on results and where are bottlenecks). just to make sure it's clear - if in a representative cases we'll see that wrapping individual objects performs better I will change the implementation to do that. I'm just not so sure that would be the case when we start distributing and updating objects in parallel which simulates real system with multiple users.
the existing test lacks parallelism, not multiple objects (manifestwork), no multiple updates in parallel (cause all pods are mapped to a single manifestwork so updates are serial), so this test is completely serial instead of parallel.

having said the above, extreme cases **SHOULD WORK** and not stop distributing (this is where fragmentation will come in).
so I think the ideal is - optimization should be oriented to representative case, but extreme cases should also work even if not with optimized performance.
let's keep syncing on this topic to identify bottlenecks and decide on the required optimizations.

**Comment by ezrasilvera** [comment]
@nirrozenbaum (cc: @dumb0002 @eaepstein )  Without getting into the two architecture options and the pros/cons, I think there are some alarming points here , some might even be considered as blockers that MUST be fixed (there are several different ways to fix them). 

1. The latency degradation itself is very problematic, while we know the reason it is still considered as unacceptable behavior in terms of user expectation.  Users expect to get the same latency/response behavior regardless of the system load  or scale, this is actually the definition of the scalability challenge.
   * Note, that putting each object in a different manifest may have similar impact on other scenarios
3. KS should never stop syncing. We should publish official limits (number of object, objects per manifest, etc.. ) and implement an active limit function that block+warn the user. We can then review these limits and decide whether they are reasonable for most of our use cases or not. 
BTW, if @dumb0002 would have used larger objects the linear slope would have been even larger and the syncing would have stopped much sooner. 

@nirrozenbaum I think it's time to pause and prepare the comparison table between the options and list all these scenarios. As your analysis shows the latency increase wasn't really a surprise and we could have anticipated  such behavior  and put it in the table before @dumb0002 even tested it (we could then conducted tests just to measure the actual numbers) .

**Comment by eaepstein** [comment]
> > @nirrozenbaum A single binding policy could be in place for multiple users on one WDS, each in their own namespace. The degradation is pretty bad even with low numbers, and appwrapper objects could be much bigger than pods. It would interesting to find a realistic scenario for which bundling per binding policy has significantly better performance in some metric: latency, memory, or CPU.
> 
> @eaepstein I didn't understand this comment. `BindingPolicy is a cluster-scoped resource. what do you mean when you say single binding policy for multiple users on one WDS, each in their own namespace? BindingPolicy isn't in a namespace.. can you please explain?

Right, the single binding policy will apply to all namespaces in the WDS. So assuming all of the user created objects have policy matching labels, they would be placed in the same bundle.

**Comment by nirrozenbaum** [comment]
> @nirrozenbaum (cc: @dumb0002 @eaepstein ) Without getting into the two architecture options and the pros/cons, I think there are some alarming points here , some might even be considered as blockers that MUST be fixed (there are several different ways to fix them).
> 
> 1. The latency degradation itself is very problematic, while we know the reason it is still considered as unacceptable behavior in terms of user expectation.  Users expect to get the same latency/response behavior regardless of the system load  or scale, this is actually the definition of the scalability challenge.
>    
>    * Note, that putting each object in a different manifest may have similar impact on other scenarios
> 2. KS should never stop syncing. We should publish official limits (number of object, objects per manifest, etc.. ) and implement an active limit function that block+warn the user. We can then review these limits and decide whether they are reasonable for most of our use cases or not.
>    BTW, if @dumb0002 would have used larger objects the linear slope would have been even larger and the syncing would have stopped much sooner.
> 
> @nirrozenbaum I think it's time to pause and prepare the comparison table between the options and list all these scenarios. As your analysis shows the latency increase wasn't really a surprise and we could have anticipated such behavior and put it in the table before @dumb0002 even tested it (we could then conducted tests just to measure the actual numbers) .

@ezrasilvera I completely disagree on point (1). no user expects to get the same latency if they distribute manifestwork with one object or manifestwork with 20 objects. the size of the object is (almost) 20 times bigger and it's obvious that it takes more time to sync it.

**Comment by nirrozenbaum** [comment]
> > > @nirrozenbaum A single binding policy could be in place for multiple users on one WDS, each in their own namespace. The degradation is pretty bad even with low numbers, and appwrapper objects could be much bigger than pods. It would interesting to find a realistic scenario for which bundling per binding policy has significantly better performance in some metric: latency, memory, or CPU.
> > 
> > 
> > @eaepstein I didn't understand this comment. `BindingPolicy is a cluster-scoped resource. what do you mean when you say single binding policy for multiple users on one WDS, each in their own namespace? BindingPolicy isn't in a namespace.. can you please explain?
> 
> Right, the single binding policy will apply to all namespaces in the WDS. So assuming all of the user created objects have policy matching labels, they would be placed in the same bundle.

you assume that all objects created by a user have the same label? and that label is the one that is used in policy to select the objects? or user is using multiple label selectors with the purpose of selecting all his created objects?
if that's the case - then of course, all will be mapped to the same bundle.. 
if that user has a single application he wants to run that might include 10-20 objects it might even make sense.
but in other cases, why would a user do that?
why would a user put the same label on all objects or select all his objects if that's not related to the same application?
again, this is not prohibited in k8s. but I think there is a misunderstanding of what's the purpose of labels in k8s.

users of k8s do not label all objects with the same label because that's missing the whole point.
the most common use case for using k8s labels is for **Grouping and Organizing - 
Labels can be used to group related resources together. For example, you can label all backend services with `app=backend` and all frontend services with `app=frontend`.
there are additional use cases for labels like:
**Identification and Selection**: Labels are used by selectors to identify and select resources. For instance, a Deployment may use a selector to specify which Pods it should manage based on their labels.

**Service Discovery**: Labels are often used in conjunction with Services for service discovery. Services can route traffic to Pods based on labels and selectors.

**Configuration and Policies**: Labels can be used to apply configurations and policies to specific sets of resources. For example, you can apply resource quotas or network policies based on labels.

none of the use cases includes labeling all objects with the same label (unless it's all related as part of a single application, which is usually not a lot of objects).. that doesn't really make sense.

**Comment by ezrasilvera** [comment]
> @ezrasilvera I completely disagree on point (1). no user expects to get the same latency if they distribute manifestwork with one object or manifestwork with 20 objects. the size of the object is (almost) 20 times bigger and it's obvious that it takes more time to sync it.

That's not the case at hand.  The user didn't create 20 Pods and tried to sync them. In that case your statement is correct . However, I thought the user was coming and was just deploying the 20th Pod as a single Pod , while using the existing ManifestWork. 
In that case the user doesn't know the internal implementation details. He just observed that few our ago when he deployed a Pod it took X sec and now when he is deploying a Pod (which may be the 100th Pod)  it takes much longer.

**Comment by nirrozenbaum** [comment]
do you assume that deploying a new pod will be mapped to the same binding policy as other pods? why?
again, anyone can come up with many unreasonable use cases. 
do you any k8s user trying to map all pods to be managed by a single deployment (e.g., deployment managing 400-500 pods)?
OR
do you see different deployment objects where each is managing several pods (for example 10) and each deployment is managing a DIFFERENT set of pods?

we all know the answer :)

that's exactly the same with BindingPolicy, it makes no sense to map a single policy to 400 pods or more. 
I don't know why we keep raising/analyzing use cases that make no sense.

have you ever tried to create a single deployment in k8s to manage few hundreds of pods in a cluster?

**Comment by nirrozenbaum** [comment]
@dumb0002 may I ask you to run a very similar test with simple change and measure the times? (if possible with graph as in your previous test).
can you try to run the same test BUT change the creation order?
for example - 
first, create 100 pods all with the same label. even though I don't think it sounds reasonable to do that, but just to compare apples to apples.
then, create the BindingPolicy to distribute all pods to a WEC.
let's measure the time it takes to distribute 100 pods in the two options (ks 0.20 and ks0.21) from the moment we create the BindingPolicy until all 100 pods have completed.

just to set the expectations in terms of analysis - 
in ks0.20 - a burst of 100 MW objects will be created (Klusterlet will need to sync 100 MW to WEC).
in ks0.21 - one MW will be created.

**Comment by nirrozenbaum** [comment]
btw, I’d be happy to hear your opinions, both options are possible but what do you think makes more sense -
1. create workload objects with some labels and then create BindingPolicy to select those objects.
OR
2. create BindingPolicy that selects nothing using labels that doesn’t exist (so selection is empty) and then start adding workload objects with the requested labels and the selection is updated in each addition.

**Comment by effi-ofer** [comment]
Rather than have manifestwork contain either 1 object or all objects, another option is to introduce a limit on the number of embedded objects in manifestwork.  We can even introduce an optimization where on initial creation the manifestwork contains all matching objects even if greater than the limit.

**Comment by nirrozenbaum** [comment]
> having said the above, extreme cases SHOULD WORK and not stop distributing (this is where fragmentation will come in).

@effi-ofer exactly. this is what I was mentioning in one of the previous messages.
fragmentation is splitting MW when it reaches a limit (not exceeding a pre-defined size limit).
that's not implemented yet. but it should be added.

**Comment by dumb0002** [comment]
> @dumb0002 may I ask you to run a very similar test with simple change and measure the times? (if possible with graph as in your previous test). can you try to run the same test BUT change the creation order? for example - first, create 100 pods all with the same label. even though I don't think it sounds reasonable to do that, but just to compare apples to apples. then, create the BindingPolicy to distribute all pods to a WEC. let's measure the time it takes to distribute 100 pods in the two options (ks 0.20 and ks0.21) from the moment we create the BindingPolicy until all 100 pods have completed.
> 
> just to set the expectations in terms of analysis - in ks0.20 - a burst of 100 MW objects will be created (Klusterlet will need to sync 100 MW to WEC). in ks0.21 - one MW will be created.

@nirrozenbaum, sure! I will conduct this experiment today

**Comment by dumb0002** [comment]
> btw, I’d be happy to hear your opinions, both options are possible but what do you think makes more sense -
> 
> 1. create workload objects with some labels and then create BindingPolicy to select those objects.
>    OR
> 2. create BindingPolicy that selects nothing using labels that doesn’t exist (so selection is empty) and then start adding workload objects with the requested labels and the selection is updated in each addition.

@nirrozenbaum, option 1 makes more sense to me to measure the time will take to distribute a burst of 100 pods all with the same label.

**Comment by nirrozenbaum** [comment]
I don’t agree with most of your claims.
there is a best practice for working with k8s and more specifically in multi cluster management. there are some leading projects in this area and there is not even one that uses the pattern you’re suggesting. all of them are wrapping multiple objects that have a relation between them (in KS this is the workload). same definition exists in SIG multi cluster.
the pattern you suggested is indeed naive. 

I didn’t say we need to tell users to create workloads first and then binding policy. what I do say is that there is a big difference in performance that depends on the creation order.

if we first create Bindingpolicy and then 100 pods we will see the attached graph above.
on the other hand, if we first create 100 pods and then the bindingpolicy we will see a different graph (I’m waiting patiently until @dumb0002 will upload it).
so now the question- which of these use cases should we optimize for?

**Comment by nirrozenbaum** [comment]
@eaepstein btw, the naive suggestion completely ignores basic aspects of k8s, for example RBAC.
in a typical k8s cluster, user A has access to namespace nsA and user B has access to nsB. each user creates his objects in his dedicated ns. the naive approach suggesting one BindingPolicy for WEC, so every object that is mapped to that WEC is selected by the same BindingPolicy. 
so now the questions - who creates this BindingPolicy? who is the owner of the object? who can remove it? user A? user B? user admin? 
and a security question from a user point of view, if you're user A - do you want to share a BindingPolicy with user B? 
in short, the answer is no. from the same reason they don't share a namespace on the hub. users that don't share permissions on the hub should **NOT** share BindingPolicy. there are many security issues here.

there are more k8s basic aspects that this suggestion is ignoring but I'll stop here.
there are some wrong assumptions along the way that are made due to the fact that we as developers work with an admin account. this is not the way users work in k8s eco system. users in any k8s system work with RBAC. we've not implemented (yet) any RBAC mechanism in KS, but that doesn't mean we can ignore it.

**Comment by dumb0002** [comment]
@nirrozenbaum  (cc: @eaepstein ), here are the plots for time it takes to distribute a burst of 100 pods from the moment we create the BindingPolicy until all 100 pods have completed. I used the following equations to compute the latency: 

```
 [t_pod-wec-completion - t_bindingPolicy-creation] - alpha

where alpha = t_pod-wec-completion - t_pod-wec-creation  (this parameter removes the latency of the pod execution time in the WEC)
```

![pod_distribuiton_to_wec_latency_ks020-v2](https://github.com/kubestellar/kubestellar/assets/25727844/6edf5887-5a23-401c-9cd7-318cb8ae2beb)


![pod_distribuiton_to_wec_latency_ks021](https://github.com/kubestellar/kubestellar/assets/25727844/c858fa50-6b25-460c-8622-90f13333564c)

For burst workload the pod distribution is better for ks-0.21

**Comment by eaepstein** [comment]
> @eaepstein btw, the naive suggestion completely ignores basic aspects of k8s, for example RBAC. in a typical k8s cluster, user A has access to namespace nsA and user B has access to nsB. each user creates his objects in his dedicated ns. the naive approach suggesting one BindingPolicy for WEC, so every object that is mapped to that WEC is selected by the same BindingPolicy. so now the questions - who creates this BindingPolicy? who is the owner of the object? who can remove it? user A? user B? user admin? and a security question from a user point of view, if you're user A - do you want to share a BindingPolicy with user B? in short, the answer is no. from the same reason they don't share a namespace on the hub. users that don't share permissions on the hub should **NOT** share BindingPolicy. there are many security issues here.

Aren't binding policies cluster scoped? If so, with an RBAC that blocks normal users from creating cluster scoped objects neither user A nor user B would be allowed to create them.

**Comment by nirrozenbaum** [comment]
> @nirrozenbaum (cc: @eaepstein ), here are the plots for time it takes to distribute a burst of 100 pods from the moment we create the BindingPolicy until all 100 pods have completed. I used the following equations to compute the latency:
> 
> ```
>  L = [t_pod-wec-completion - t_bindingPolicy-creation] - alpha
> 
> where alpha = t_pod-wec-completion - t_pod-wec-creation  (this parameter removes the latency of the pod execution time in the WEC)
> ```
> 
> ![pod_distribuiton_to_wec_latency_ks020-v2](https://private-user-images.githubusercontent.com/25727844/317925632-6edf5887-5a23-401c-9cd7-318cb8ae2beb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE3MDUwMDcsIm5iZiI6MTcxMTcwNDcwNywicGF0aCI6Ii8yNTcyNzg0NC8zMTc5MjU2MzItNmVkZjU4ODctNWEyMy00MDFjLTljZDctMzE4Y2I4YWUyYmViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI5VDA5MzE0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzZDlmNmQ5NTBjZTY0MDVmMjE5NGE4MzYxZWY1NzA3ZDg3ZTEyODZkMjkzOTI2MDY0MjE2ODQ4NTQzY2U3YzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MUmBdDzwgZDyIxsn0bS5Lq3B5D-MQbwQdJRa8-qgGTA)
> 
> ![pod_distribuiton_to_wec_latency_ks021](https://private-user-images.githubusercontent.com/25727844/317925710-c858fa50-6b25-460c-8622-90f13333564c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE3MDUwMDcsIm5iZiI6MTcxMTcwNDcwNywicGF0aCI6Ii8yNTcyNzg0NC8zMTc5MjU3MTAtYzg1OGZhNTAtNmIyNS00NjBjLTg2MjItOTBmMTMzMzM1NjRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI5VDA5MzE0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxMGVlYzZmNTE5NjFhNmE0ZWNiZmIzMjYwMDc0ZGFkYjE1MmZiMDFmZDkwZDEwNjIzNmM4YWZlNTk5NDkzNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kCJjgpDH1NnhpKKkaW5BFydrVscfnGFCUV6x8Rn4ZeI)
> 
> For burst workload the pod distribution is better for ks-0.21

@dumb0002 many thanks for conducting this experiment.
so what we see here is the following - 
when creating first the workload objects and then the BindingPolicy in this experiment, the time to completion using ks0.21 was 32 seconds vs 2:10 minutes us ks0.20.
on the other hand when creating first the binding policy and then 100 workload objects, the 100th pod completion time take 23-24 second in ks0.20 vs 1:09 in ks0.21.
this experiment is nice and shows differences in performance depends on creation order. 
clarification - I don't think we should recommend any creation order, users should create in any order they want.

even though this experiment is nice, this is a serial test (everything is done serially), so this is very far away from reality in most cases. 
to have a good test for KS, we should design and conduct experiments that use parallelism with some randomness.
e.g., in some cases add/update/delete objects in existing BindingPolicy (which in turn updates existing MW), in other cases create new objects and distribute them using new BP, in some cases create objects first then BP and in others switch the creation order. all of this should happen in parallel by multiple threads that will simulate multiple users.
and of course, we need to define what are the metrics we want to compare.

**Comment by nirrozenbaum** [comment]
> Aren't binding policies cluster scoped? If so, with an RBAC that blocks normal users from creating cluster scoped objects neither user A nor user B would be allowed to create them.

@eaepstein if I it right, you suggest that KS users won't have access to BindingPolicy and it's a system auto-created object per WEC.

your suggestion ignores RBAC and more specifically users isolation, it ignores grouping and organization of WECs which is a basic aspect in large scale (it ignores scale in general), it ignores status summarization and status in general, it ignores the option for dynamic changes like dynamic customization and filtering, and more..
By design, BindingPolicy is a user facing API. that's the entry point for users to define things dynamically including how to summary status, how to dynamically customize and filter objects (see for example the latest requirement from Argo integration here - #1976). 

I invite you to raise this suggestion for the broader team (slack/poll/webex.. doesn't matter) and see their responses. 
I doubt this is the model we want to sell for KS.

**Comment by pdettori** [comment]
I have been looking into several use cases for KubeStellar and integrations with other projects such as Argo Workflows, MCAD, Kueue and KubeFlow Pipelines. None of these use cases would require hundreds or thousands of clusters, but rather a handful of them. On the other hand, objects to deliver could be quite large (an Argo Workflow object may contain python code, plenty of metadata and lots of pod specs). In most of these integrations, the simplest approach is to create one policy per cluster (typically, each workflow or job would be delivered just to one cluster) and rely simply on labels or other identifiers in the objects to route them to the selected target cluster. I think that one of the key goals of KubeStellar is to reduce friction for adopting multi-cluster, and it seems to me that putting the burden on the user to create for example one binding policy for each workload is asking too much. This is something we saw also with the MCAD integration, which was drastically simplified just labeling the MCAD objects. My recommendation is to a) evaluate performances with N larger workload objects per binding policy (I can provide argo workflows sample generated by KFP) b) work on a design for ManifestWork sharding based on ManifestWork size, where  max size per shard is tunable so that the it can adapt the different use cases.

**Comment by ezrasilvera** [comment]
@pdettori I agree that there are many use cases that just want WorkLoad -->single WEC (and then the main problem is scheduling algorithm).  However I don't want us to be biased from the (great) recent  integration work and design KS as if that's the main use case. 
Remember that we started with use cases aimed for Edge and maybe "branch offices"  in which the common patter is WL --> many WECs.  
BTW, even on your integration scenarios, if we also want to deploy the base components through KS (not just the workload) like - monitoring agents, agents etc.. -  that deployment is 1:N and not 1:1

**Comment by pdettori** [comment]
@ezrasilvera I was just presenting use cases that are of immediate applicability for us. But independently from 1:1 or 1:N, the problem of a too large manifest could present itself, therefore we should be able to accomodate these different use cases.

**Comment by dumb0002** [comment]
> any thanks for conducting this experiment. so what we see here is the following - when creating first the workload objects and then the BindingPolicy in this experiment, the time to completion using ks0.21 was 32 seconds vs 2:10 minutes using ks0.20. on the other hand when creating first the binding policy and then 100 workload objects, the 100th pod completion time take 23-24 second in ks0.20 vs 1:09 in ks0.21. this experiment is nice and shows differences in performance depends on creation order. clarification - I don't think we should recommend any creation order, users should create in any order they want.
> 
> even though this experiment is nice, this is a serial test (everything is done serially), so this is very far away from reality in most cases.

@nirrozenbaum, the plots posted at the beginning of this issue is the E2E latency using a **sequential function** to generate the workload in a K3D cluster. Hence, for accurate comparison, I repeated the experiment for creating first the binding policy and then a burst of 100 workload objects (there is **parallelism** in this burst workload generation): 

To compute latency to distributed the pods to the WEC are used the following equation: 

```
L = [t_pod-wec-completion - t_pod-wds1-creation] - alpha

where alpha = t_pod-wec-completion - t_pod-wec-creation  (this parameter removes the latency of the pod execution time in the WEC)
```

![pod_distribuiton_to_wec_latency_ks020-case1](https://github.com/kubestellar/kubestellar/assets/25727844/059ffe79-94f3-43ff-b426-a8313355de80)


![pod_distribuiton_to_wec_latency_ks021-case1](https://github.com/kubestellar/kubestellar/assets/25727844/a34e18f1-99dc-47b5-89eb-a8a2501b62cf)


Comparing these plots we can observe that  ks0.21 performs better for **burst** workloads and Yes there is differences in performance depends on creation order of BindingPolicy & workload: 

   a)  when creating first the workload objects and then the BindingPolicy in this experiment, the time to completion using ks0.21 was 33 seconds vs 2:10 minutes using ks0.20

   b)  when creating first the binding policy and then 100 workload objects, the time to completion using ks0.21 was 32 seconds vs 1:20 minutes using ks0.20.

I'll design some experiments with some degree of parallelism and randomness to discuss with the team so that we can all come to an agreement of a good representative workload and traffic patterns.

**Comment by dumb0002** [comment]
> I have been looking into several use cases for KubeStellar and integrations with other projects such as Argo Workflows, MCAD, Kueue and KubeFlow Pipelines. None of these use cases would require hundreds or thousands of clusters, but rather a handful of them. On the other hand, objects to deliver could be quite large (an Argo Workflow object may contain python code, plenty of metadata and lots of pod specs). In most of these integrations, the simplest approach is to create one policy per cluster (typically, each workflow or job would be delivered just to one cluster) and rely simply on labels or other identifiers in the objects to route them to the selected target cluster. I think that one of the key goals of KubeStellar is to reduce friction for adopting multi-cluster, and it seems to me that putting the burden on the user to create for example one binding policy for each workload is asking too much. This is something we saw also with the MCAD integration, which was drastically simplified just labeling the MCAD objects. My recommendation is to a) evaluate performances with N larger workload objects per binding policy (I can provide argo workflows sample generated by KFP) b) work on a design for ManifestWork sharding based on ManifestWork size, where max size per shard is tunable so that the it can adapt the different use cases.

@pdettori, agreed, for KubeStellar and integrations with other upper level application, the user shouldn't be burden to create binding policies, instead it can rely simply on labels or other identifiers to route objects to the target remote cluster. That's what we're doing right now in our MCC stack. 

Could you provide the sample  argo workflows generated by KFP that you mentioned above?

**Comment by nirrozenbaum** [comment]
while things may make sense in one integration or another, the question here is deeper. 

the question is about KS model and the design. 
for example, if KS model for distribution is 1:1 that’s completely different than 1:N, and the design, architecture and optimizations will be made very differently depends on the model.

**Comment by eaepstein** [comment]
> while things may make sense in one integration or another, the question here is deeper.
> 
> the question is about KS model and the design. for example, if KS model for distribution is 1:1 that’s completely different than 1:N, and the design, architecture and optimizations will be made very differently depends on the model.

Right. The one architectural design that has clear impact on the usage scenarios discussed above is that BPs are currently cluster-scoped. I also agree with this design choice because it offers simpler usage scenarios and better scalability.

**Comment by eaepstein** [comment]
> To compute latency to distributed the pods to the WEC are used the following equation:
> 
> ```
> L = [t_pod-wec-completion - t_pod-wds1-creation] - alpha
> 
> where alpha = t_pod-wec-completion - t_pod-wec-creation  (this parameter removes the latency of the pod execution time in the WEC)
> ```

@dumb0002 These are surprising plots, particularly the grouping of latency values exactly on one second intervals for ks021. To disambiguate delays due to bundling+transport+klusterlet submission from delays due to K8S deploying the pods it could be interesting to see ks020 and ks021 plots of:
```
bundling+transport+klusterlet = t_pod-wec-creation - t_pod-wds1-creation
```
and
```
kubernetes deployment = t_pod-wec-completion - t_pod-wec-creation - 20
```

I notice that when running the stress test and metrics together, kubectl reports for wec1:
```
stress-pod-3865   0/1     Pending             0          0s
stress-pod-3865   0/1     Pending             0          0s
stress-pod-3865   0/1     ContainerCreating   0          0s
stress-pod-3865   1/1     Running             0          1s
stress-pod-3866   0/1     Pending             0          0s
stress-pod-3866   0/1     Pending             0          0s
stress-pod-3866   0/1     ContainerCreating   0          0s
stress-pod-3866   1/1     Running             0          1s
```
Delays from wec creating to running could be much longer in burst testing.

**Comment by nirrozenbaum** [comment]
> Right. The one architectural design that has clear impact on the usage scenarios discussed above is that BPs are currently cluster-scoped. I also agree with this design choice because it offers simpler usage scenarios and better scalability.

This wasn't what I was referring to. We all agree that BPs should be cluster-scoped. there is no other way to select cluster-scoped resources for distribution.
I was referring to the question of whether BPs are fixed, e.g., one per WEC, or dynamic, e.g., created dynamically per workload and includes grouping and organization of clusters, while each workload may include different set of clusters.
these were (some of) the design choices I was referring to

**Comment by eaepstein** [comment]
> > Right. The one architectural design that has clear impact on the usage scenarios discussed above is that BPs are currently cluster-scoped. I also agree with this design choice because it offers simpler usage scenarios and better scalability.
> 
> This wasn't what I was referring to. We all agree that BPs should be cluster-scoped. there is no other way to select cluster-scoped resources for distribution. I was referring to the question of whether BPs are fixed, e.g., one per WEC, or dynamic, e.g., created dynamically per workload and includes grouping and organization of clusters, while each workload may include different set of clusters. these were (some of) the design choices I was referring to

It could make sense to give cluster-scoped auth to a single user (or closely aligned users) sharing a WDS. Multiple uncoordinated users on the same WDS normally would not have such auth. It would be nice to have good performance for both of these scenarios.

**Comment by nirrozenbaum** [comment]
adding here information just to verify we are all aware - 
In kubernetes, it's possible to give users access only to their namespace and access to create a specific cluster-scoped object. for example, a user may have access to create anything in his namespace (admin of that namespace) and access to create BindingPolicy objects.
the option to create a specific cluster-scoped object doesn't make this user an admin or a super-user. 
this is not a general cluster-scoped auth but a very specific one.

**Comment by dumb0002** [comment]
> @dumb0002 These are surprising plots, particularly the grouping of latency values exactly on one second intervals for ks021. To disambiguate delays due to bundling+transport+klusterlet submission from delays due to K8S deploying the pods it could be interesting to see ks020 and ks021 plots of:
> 
> ```
> bundling+transport+klusterlet = t_pod-wec-creation - t_pod-wds1-creation
> ```
> 
> and
> 
> ```
> kubernetes deployment = t_pod-wec-completion - t_pod-wec-creation - 20
> ```
> 

@eaepstein, the equation that I used above is already removing  delays due to K8S deploying the pods - that's achieved by subtracting the `alpha` parameter

**Comment by dumb0002** [comment]
> I notice that when running the stress test and metrics together, kubectl reports for wec1:
> 
> ```
> stress-pod-3865   0/1     Pending             0          0s
> stress-pod-3865   0/1     Pending             0          0s
> stress-pod-3865   0/1     ContainerCreating   0          0s
> stress-pod-3865   1/1     Running             0          1s
> stress-pod-3866   0/1     Pending             0          0s
> stress-pod-3866   0/1     Pending             0          0s
> stress-pod-3866   0/1     ContainerCreating   0          0s
> stress-pod-3866   1/1     Running             0          1s
> ```
> 
> Delays from wec creating to running could be much longer in burst testing.

@eaepstein, that's expected that burst workload will have a high latency for pod creation in the wecs when compared to the sequential workload generation

**Comment by eaepstein** [comment]
> adding here information just to verify we are all aware - In kubernetes, it's possible to give users access only to their namespace and access to create a specific cluster-scoped object. for example, a user may have access to create anything in his namespace (admin of that namespace) and access to create BindingPolicy objects. the option to create a specific cluster-scoped object doesn't make this user an admin or a super-user. this is not a general cluster-scoped auth but a very specific one.

Actually no, I was not aware that namespaced-scoped users would  still be able to create a BP. That is good to know.

**Comment by MikeSpreitzer** [comment]
I am puzzled by one thing in the data shown in the opening comment. My recollection of the reports directly from @dumb0002 for KS 0.20 (or was it KS 0.21?) is that WDS-to-WEC latency is higher and more variable because the klusterlet operates periodically with a fairly large period. The reports from @dumb0002 showed step-by-step breakdowns; one of the steps was from ITS (formerly IMBS) to WEC and that shows the latency due to the klusterlet's behavior. My memory may be wrong. I do not have a pointer to such a document handy. I think it would be great to look at one of those reports.

**Comment by MikeSpreitzer** [comment]
> Does each entry in the bundled manifestwork have a create time?

Yes. I am testing release 0.21.2-rc2, and in the ManifestWork objects I see that each wrapped workload object retains its `metadata.creationTimestamp` (that is a standard bit of Kubernetes object metadata).

> Do entries also have the ID of the associated workload objects to be able to measure transport times?

Yes, it would be problematic if not. In Kubernetes an object is identified by name and namespace (among objects of the same kind), and user code knows this and, in general, works with this.

@dumb0002 has been producing reports that break down the end-to-end times into individual steps, relying on standard Kubernetes metadata.

**Comment by MikeSpreitzer** [comment]
@nirrozenbaum wrote:

> If that's ok with all of you, give me around two weeks and I'll try to come up with stress test setup that represents KS usage better.

I would differentiate between _defining_ a stress test and _implementing_ a stress test. I would be ready to discuss stress test _definitions_ without waiting for implementation. After we get our current bugs out and customization done, of course.

**Comment by MikeSpreitzer** [comment]
> if in a representative cases we'll see that wrapping individual objects performs better I will change the implementation to do that.

How do we really know what are representative cases? We are designing for a range of possible uses. Remember edge comuting? The MCAD stack is not that. Do we have to pick just one? I think that we are in too early a stage to rule out use cases that we have considered. OTOH, I think it is fine to focus our current work on the use cases that have the most immediate traction.

It seems to me that if we find that we want to support a mixture of use cases, in which some work best with bundling and some work best with individual object wrapping, then this is something we could definitely do. I suspect that in such a world it would be known at deployment time which is the better way to operate; in that case the generic transport controller code could easily take a command line flag telling it which way to go.

**Comment by MikeSpreitzer** [comment]
>  Users expect to get the same latency/response behavior regardless of the system load or scale.

Users expect ponies too. I think we have to settle for satisfying reasonably educated users.

**Comment by MikeSpreitzer** [comment]
I am running out of time, have not read all the comments yet. I will leave a few thoughts of mine on usage scenarios.

I think the orientation of multiple "users" sharing a WDS is not very compelling. It reminds me of time-sharing. I am more drawn to thinking about how an organization uses a larger system that includes KubeStellar. For example, in edge computing scenarios. There the number of WECs is very large. And with a very large number of WECs, it is practically guaranteed that there will always be some coming and going. The rate of change in the population of WECs will likely be greater than the rate of change in the workload and BidingPolicies. In such a scenario, KubeSteller is not being used by a bunch of independent users. It is used by a cooperating collection of teams collaborating on managing what is going on at the edge.

Another sort of usage scenario is 1:1 dispatching of some sort of work among multiple execution clusters. The MCAD stack is an extremely small and simple version of this. Some of us remember working with the Octopus team on a little more general version. Some higher level thing occasionally drops another workflow into a WDS. Some other central controller factors the workflow into pieces and decides which execution cluster each piece should go to. An execution cluster can potentially handle multiple workflow pieces at once, depending on the cluster's capacity and the needs of the pieces. In these scenarios, there can be one standing BindingPolicy for each execution cluster and the dispatched work flows through the KubeStellar pipes to the WECs. To get down to numbers worth calculating with, you have to get into application specifics. An ML training job is going to be so big that it will entirely consume a WEC. I am less clear on the data prep steps, but they are going to be pretty big too. OTOH, individual inferencing jobs may to too small to make sense for dispatching via KubeStellar. Don't look to me for numbers here, ask people who know. Don't only look at ML.

**Comment by MikeSpreitzer** [comment]
Also: KubeStellar is _not_ "transparent multi-cluster". It is intentional and explicit management of workload in remote clusters.

**Comment by nirrozenbaum** [comment]
> @nirrozenbaum wrote:
> 
> > If that's ok with all of you, give me around two weeks and I'll try to come up with stress test setup that represents KS usage better.
> 
> I would differentiate between _defining_ a stress test and _implementing_ a stress test. I would be willing to discuss stress test _definitions_ without waiting for implementations. After we get our current bugs out and customization done, of course.

@MikeSpreitzer I agree. we can try to find some time to define stress test after current work is completed.

**Comment by clubanderson** [comment]
Issues go stale after 90d of inactivity.
After a furter 30 days, they will turn rotten.
Mark the issue as fresh with `/remove-lifecycle stale`.

If this issue is safe to close now please do so with `/close`.

/lifecycle stale

**Comment by clubanderson** [comment]
Stale issues rot after 30d of inactivity.
Mark the issue as fresh with `/remove-lifecycle rotten`.
Rotten issues close after an additional 30d of inactivity.

If this issue is safe to close now please do so with `/close`.

/lifecycle rotten

**Comment by clubanderson** [comment]
Rotten issues close after 30d of inactivity.
Reopen the issue with `/reopen`.
Mark the issue as fresh with `/remove-lifecycle rotten`.

/close
